corehq/messaging/scheduling/scheduling_partitioned/models.py

```diff
@@ -0,0 +1,203 @@
+
+        return obj
+
+    def expand_recipients(self):
+        """
+        Can be used as a generator to iterate over all individual contacts who
+        are the recipients of this ScheduleInstance.
+        """
+        if self.recipient_type in ('CommCareCase', 'CommCareUser', 'WebUser'):
+            yield self.recipient
+        elif self.recipient_type == 'CommCareCaseGroup':
+            case_group = self.recipient
+            for case in case_group.get_cases():
+                yield case
+        elif self.recipient_type == 'Group':
+            group = self.recipient
+            for user in group.get_users(is_active=True, only_commcare=False):
+                yield user
+        elif self.recipient_type == 'Location':
+            location = self.recipient
+            if self.memoized_schedule.include_descendant_locations:
+                location_ids = location.get_descendants(include_self=True).filter(is_archived=False).location_ids()
+            else:
+                location_ids = [location.location_id]
+
+            user_ids = set()
+            for location_id in location_ids:
+                for user in get_all_users_by_location(self.domain, location_id):
+                    if user.is_active and user.get_id not in user_ids:
+                        user_ids.add(user.get_id)
+                        yield user
+        else:
+            raise UnknownRecipientType(self.recipient_type)
+
+    def handle_current_event(self):
+        content = self.memoized_schedule.get_current_event_content(self)
+        for recipient in self.expand_recipients():
+            content.send(recipient)
+        # As a precaution, always explicitly move to the next event after processing the current
+        # event to prevent ever getting stuck on the current event.
+        self.memoized_schedule.move_to_next_event(self)
+        self.memoized_schedule.move_to_next_event_not_in_the_past(self)
+
+    @property
+    def schedule(self):
+        raise NotImplementedError()
+
+    @schedule.setter
+    def schedule(self, value):
+        raise NotImplementedError()
+
+    @property
+    @memoized
+    def memoized_schedule(self):
+        """
+        This is named with a memoized_ prefix to be clear that it should only be used
+        when the schedule is not changing.
+        """
+        return self.schedule
+
+
+class AlertScheduleInstance(ScheduleInstance):
+    alert_schedule_id = models.UUIDField()
+
+    class Meta(ScheduleInstance.Meta):
+        db_table = 'scheduling_alertscheduleinstance'
+
+    @property
+    def schedule(self):
+        return AlertSchedule.objects.get(schedule_id=self.alert_schedule_id)
+
+    @schedule.setter
+    def schedule(self, value):
+        if not isinstance(value, AlertSchedule):
+            raise ValueError("Expected an instance of AlertSchedule")
+
+        self.alert_schedule_id = value.schedule_id
+
+
+class TimedScheduleInstance(ScheduleInstance):
```

Comment:
yeah these are the only two that we are sharding so far. they do have foreign keys to objects in the main db (each schedule instance model has a pointer to its schedule definition), but those are not handled with a django ForeignKey to avoid issues with querying / joining. those foreign keys are defined here and are only traversable through a python property:

https://github.com/dimagi/commcare-hq/blob/4997ea281f2b9c2f0b86c5e278b8a030a7e10a4b/corehq/messaging/scheduling/scheduling_partitioned/models.py#L160

https://github.com/dimagi/commcare-hq/blob/4997ea281f2b9c2f0b86c5e278b8a030a7e10a4b/corehq/messaging/scheduling/scheduling_partitioned/models.py#L178